File: examples/hello_pubsub/hello_pubsub_cpp/src/hello_subscriber.cpp
/**
 *  A minimal ROS2 cpp node that subscribes to a topic.
 *
 *  Basic structure for ROS2 node:
 *  - Initialise rclcpp
 *  - Create a node
 *  - Create the subscriber
 *  - Keep it alive with spin
 *  - Shutdown
 *
 *  BEST_EFFORT and VOLATILE have been selected for the QOS_PROFILE, as the
 *  publisher uses them.
 *
 *  Author: Ioannis Selinis <selinis.g@gmail.com> 2025
 */

#include <rclcpp/rclcpp.hpp>
#include "example_interfaces/msg/string.hpp"


static const rmw_qos_profile_t rmw_qos_profile =
{
    RMW_QOS_POLICY_HISTORY_KEEP_LAST,
    1,
    RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT,
    RMW_QOS_POLICY_DURABILITY_VOLATILE,
    RMW_QOS_DEADLINE_DEFAULT,
    RMW_QOS_LIFESPAN_DEFAULT,
    RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
    RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
    false
};


class HelloSubscriber : public rclcpp::Node
{
    public:
        HelloSubscriber() : Node("hello_c_subscriber")
        {
            auto qos = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile), rmw_qos_profile);
            m_subscriber = this->create_subscription<example_interfaces::msg::String>("hello_world",
                                                                                      qos,
                                                                                      std::bind(&HelloSubscriber::topicCallback, this, std::placeholders::_1));
            RCLCPP_INFO(this->get_logger(), "Node has started");
        }

    private:
        void topicCallback(const example_interfaces::msg::String::SharedPtr msg)
        {
            RCLCPP_INFO(this->get_logger(), "Received %s", msg->data.c_str());
        }
        rclcpp::Subscription<example_interfaces::msg::String>::SharedPtr m_subscriber;
};


int main (int argc, char **argv)
{
    rclcpp::init(argc, argv);                        // Initialise rclcpp (ROS communications)
    auto node = std::make_shared<HelloSubscriber>(); // Smart/Shared pointer
    rclcpp::spin(node);                              // Keep node alive
    rclcpp::shutdown();                              // Shutdown executor
}

File: examples/hello_pubsub/hello_pubsub_py/hello_pubsub_py/hello_subscriber.py
#!/usr/bin/env python3

'''
    A minimal ROS2 python node that subscribes to a topic.

    Basic structure for ROS2 node publisher:
    - Initialise rclpy
    - Create a node
    - Create the subscriber
    - Keep it alive with spin
    - Shutdown

    Since the publisher node uses BEST_EFFORT and VOLATILE, the subscriber
    has to use the same policies.

    Author: Ioannis Selinis <selinis.g@gmail.com> 2025
'''


import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
from example_interfaces.msg import String


# https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html
QOS_PROFILE = QoSProfile(
    reliability=ReliabilityPolicy.BEST_EFFORT,      # RELIABLE or BEST_EFFORT
    durability=DurabilityPolicy.VOLATILE,           # VOLATILE or TRANSIENT_LOCAL
    history=HistoryPolicy.KEEP_LAST,                # KEEP_ALL or KEEP_LAST
    depth=1                                         # Used only with KEEP_LAST
)


class HelloSubscriber(Node):
    """The HelloSubscriber class that inherits from rclpy.node.Node.
    Initialises with a subscriber.
    """
    def __init__(self):
        super().__init__(node_name='hello_subscriber')

        self.subscriber = self.create_subscription(msg_type=String,
                                                   topic='hello_world',
                                                   callback=self.hello_world_callback,
                                                   qos_profile=QOS_PROFILE)


    def hello_world_callback(self, msg: String) -> None:
        """The callback of the hello_world topic

        Keyword arguments:
        msg (String) -- The ROS2 message
        """
        self.get_logger().info(f'Received message: {msg.data}')


def main(args=None) -> None:
    """Entry Point for our node"""
    rclpy.init(args=args)   # Initialise rclpy (ROS communications)
    node = HelloSubscriber()
    rclpy.spin(node=node)   # Keep node alive
    node.destroy_node()     # Destroy node
    rclpy.shutdown()        # Shutdown the executor


if __name__ == '__main__':
    main()
